### PR TITLE
fs: add merge partitions support

### PR DESCRIPTION
**Pull request by gneworld**
## Summary

Usage example for merging the factory partition and reserve partition into the merge partition:
register_merge_blockdriver("/dev/merge", "/dev/factory", "/dev/reserve", NULL)

## Impact
FS

## Testing

test on bes platform

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No, this PR does not fully meet the NuttX requirements as described.  While it provides a usage example and mentions testing, it lacks crucial information.

Here's a breakdown of missing information:

* **Summary:**  Needs to explain *why* this change is necessary. Is it a new feature?  Fixing a bug?  Improving performance?  What problem does it solve?  It also lacks any mention of related issues.
* **Impact:**  While it mentions "FS" (presumably filesystem), it needs to be far more specific.  It needs to answer all the yes/no questions about user impact, build impact, hardware impact, documentation needs, security implications, and compatibility.  Simply saying "FS" isn't helpful.
* **Testing:**  "test on bes platform" is insufficient.  It needs to specify the host OS, CPU, compiler used for building, and the specific target architecture, board, and configuration.  Most importantly, it lacks the required "before" and "after" testing logs.  Without these logs, there's no evidence provided that the change actually works or improves anything.


In short, the PR needs significantly more detail to be considered complete.

File: fs/driver/fs_blockmerge.c [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+ ****************************************************************************/
+
+static int merge_open(FAR struct inode *inode)
+{
+  FAR struct merge_priv_s *priv;
+  int ret = 0;
+  size_t i;
+
+  DEBUGASSERT(inode && inode->i_private);
+  priv = (FAR struct merge_priv_s *)inode->i_private;
```
Comment (anchao) [inline fs/driver/fs_blockmerge.c:96]:
```suggestion
  priv = inode->i_private;
```

File: fs/driver/fs_blockmerge.c [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+  return ret;
+}
+
+static int merge_close(FAR struct inode *inode)
+{
+  FAR struct merge_priv_s *priv;
+  size_t i;
+
+  DEBUGASSERT(inode && inode->i_private);
+  priv = (FAR struct merge_priv_s *)inode->i_private;
```
Comment (anchao) [inline fs/driver/fs_blockmerge.c:152]:
```suggestion
  priv = inode->i_private;
```

File: fs/driver/fs_blockmerge.c [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+static ssize_t merge_read(FAR struct inode *inode, unsigned char *buffer,
+                          blkcnt_t startsector, unsigned int nsectors)
+{
+  FAR struct merge_priv_s *priv;
+  ssize_t nread = 0;
+  size_t ntotal = 0;
+  size_t i;
+
+  DEBUGASSERT(inode && inode->i_private);
+  priv = (FAR struct merge_priv_s *)inode->i_private;
```
Comment (anchao) [inline fs/driver/fs_blockmerge.c:177]:
```suggestion
  priv = inode->i_private;
```

File: fs/driver/fs_blockmerge.c [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+      ntotal += nwritten;
+      startsector = 0;
+    }
+
+  /* On success, return the number of blocks written */
+
+  return ntotal ? ntotal : nwritten;
+}
+
+static int merge_geometry(FAR struct inode *inode, struct geometry *geometry)
```
Comment (anchao) [inline fs/driver/fs_blockmerge.c:246]:
```suggestion
static int merge_geometry(FAR struct inode *inode, FAR struct geometry *geometry)
```

File: fs/driver/fs_blockmerge.c [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+      startsector = 0;
+    }
+
+  /* On success, return the number of blocks read */
+
+  return ntotal ? ntotal : nread;
+}
+
+static ssize_t merge_write(FAR struct inode *inode,
+                           const unsigned char *buffer,
```
Comment (anchao) [inline fs/driver/fs_blockmerge.c:207]:
```suggestion
                           FAR const unsigned char *buffer,
```

File: fs/driver/fs_blockmerge.c [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+          ferr("ERROR: Failed to close block driver for %s\n",
+               priv->part[i].path);
+          return ret;
+        }
+    }
+
+  return OK;
+}
+
+static ssize_t merge_read(FAR struct inode *inode, unsigned char *buffer,
```
Comment (anchao) [inline fs/driver/fs_blockmerge.c:168]:
```suggestion
static ssize_t merge_read(FAR struct inode *inode, FAR unsigned char *buffer,
```

**Review by anchao — CHANGES_REQUESTED** [review]

**Review by anchao — APPROVED** [review]

**Comment by lupyuen** [comment]
Sorry I have to cancel the CI Check, there's a problem with CI Test:
- https://github.com/apache/nuttx/pull/14667#issuecomment-2461197521

**Pull request merged**